src/wrapFetch.ts

```diff
@@ -0,0 +1,86 @@
+import { useOnError, useToken } from "AuthProvider";
+import withNotification from "withNotification";
+import Errors from "./errors";
+
+type fetchFunc = (
+  input: RequestInfo,
+  init?: RequestInit | undefined
+) => Promise<Response>;
+
+interface Options {
+  notifications?: boolean;
+}
+
+export default function wrapFetch(
+  fetch: fetchFunc,
+  opts: Options = {}
+): fetchFunc {
+  if (opts.notifications) {
+    fetch = wrapNotifications(fetch);
+  }
+
+  return (input, init) => getDecodeFetch(fetch, input, init);
+}
+
+function wrapNotifications(fetch: fetchFunc): fetchFunc {
+  return (input, init) => getNotificationsFetch(fetch, input, init);
+}
+
+async function getNotificationsFetch(
+  fetch: fetchFunc,
+  input: RequestInfo,
+  init?: RequestInit | undefined
+) {
+  let onlyFail = !init || !init.method || init.method.toLowerCase() === "get";
+
+  return await withNotification(
+    () => fetch(input, init),
+    input.toString(),
+    onlyFail ? "fail" : "all"
+  );
+}
+
+async function getDecodeFetch(
+  fetch: fetchFunc,
+  input: RequestInfo,
+  init?: RequestInit | undefined
+) {
+  let token = useToken();
```

Comment:
I think this will throw an error at times? Because this (1) is not a hook (2) is not guaranteed to be inside of a component (eg, someone creates a file called `api.ts` where they put all their fetch calls)

This is the trickiest part of the implementation! How do we expose this token _outside_ of the React component tree? (eg outside of context/props)

We need some way to have global, mutable state (the token). Here are two ideas:

**1. Single instance pattern**

[Something like this](https://k94n.com/es6-modules-single-instance-pattern). This class instance simply holds on to the latest value of the token.

`AuthProvider` will write to this class instance, everything else (eg here in `wrapFetch`) will just read.

`AuthProvider` will set the token on the instance _before_ rendering the component tree, guaranteeing that the token will be there when any children go to make a fetch call via `wrapFetch`

**2. Just use localstorage**

Honestly, we're **already storing tokens here**, and it's a shared, global store. So, all you'd need to do is write a little interface between local storage and the two areas that need to read from it (`AuthProvider` and `wrapFetch`) and the one area that will write to it (`AuthProvider`)

If we go this route, then we have no choice but to always cache the token (eg remove `cacheToken`, default to always storing it there). I'm OK with this, as we can always figure out how to support non-cached tokens later. (Supporting them now is a bit of a pre-optimization)

**Note** that in either case, we still want to have the hooks like `useToken()` as exposed convenience functions for our users.

Reply:
I opted for the first option, and implemented the `AuthProviderHelper` class to hold `org`, `token` and `initialized` (to check if the app is wrapped inside `AuthProvider`), while the `AuthProvider` notifies changes using `useEffect`.